randovania/game_description/game_database_view.py

```diff
@@ -0,0 +1,304 @@
+        """
+
+    @abc.abstractmethod
+    def node_by_identifier(self, identifier: NodeIdentifier) -> Node:
+        """
+        Find a node with the given NodeIdentifier.
+        Raises KeyError if no node could be found.
+        """
+
+    @abc.abstractmethod
+    def assert_pickup_index_exists(self, index: PickupIndex) -> None:
+        """
+        If the PickupIndex does not exist, this function raises an Exception
+        """
+
+    @abc.abstractmethod
+    def create_resource_collection(self) -> ResourceCollection:
+        """
+        Creates a new ResourceCollection
+        """
+
+    @abc.abstractmethod
+    def default_starting_location(self) -> NodeIdentifier:
+        """
+        The default starting location for the game. Not really used since the preset starting location replaces this...
+        """
+
+    @abc.abstractmethod
+    def get_dock_types(self) -> list[DockType]:
+        """
+        List all available DockTypes
+        """
+
+    @abc.abstractmethod
+    def get_dock_weakness(self, dock_type_name: str, weakness_name: str) -> DockWeakness:
+        """
+        Gets a DockWeakness via names
+        """
+
+    @abc.abstractmethod
+    def get_resource_database_view(self) -> ResourceDatabaseView:
+        """
+        Gets a view for the ResourceDatabase
+        """
+
+    @abc.abstractmethod
+    def get_pickup_database(self) -> PickupDatabase:
+        """
+        Gets the PickupDatabase for this game.
+        """
+
+    @abc.abstractmethod
+    def get_victory_condition(self) -> Requirement:
+        """
+        Gets the requirement that determines if the player has won.
+        """
+
+    @abc.abstractmethod
+    def pickup_nodes_with_feature(self, feature: HintFeature) -> tuple[PickupNode, ...]:
+        """
+        Returns an iterable tuple of PickupNodes with the given feature (either directly or in their area)
+        """
+
+    @abc.abstractmethod
+    def node_from_pickup_index(self, index: PickupIndex) -> PickupNode:
+        """
+        Returns the PickupNode with the given index.
+        :raises: KeyError if it doesn't exist
+        """
+
+    @abc.abstractmethod
+    def area_from_node(self, node: Node) -> Area:
+        """
+        Returns the Area that contains the given node.
+        :raises: KeyError if it doesn't exist
+        """
+
+
+class GameDatabaseViewProxy(GameDatabaseView):
+    """
+    A GameDatabaseView, implemented by delegating all calls to another GameDatabaseView.
+    Intended to be used to overwrite specific functions.
+    """
+
+    def __init__(self, original: GameDatabaseView):
+        self._original = original
+
+    @override
+    def node_iterator(self) -> Iterable[tuple[Region, Area, Node]]:
+        return self._original.node_iterator()
+
+    @override
+    def iterate_nodes_of_type[NodeT: Node](self, node_type: type[NodeT]) -> Iterator[tuple[Region, Area, NodeT]]:
+        return self._original.iterate_nodes_of_type(node_type)
```

Comment:
I also never know whether it's better to `return` or `yield from` in cases like this

Reply:
`return` is faster so I'll prefer that.